File: app/Product.php
<?php

namespace App;

use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Input;
use Illuminate\Support\Facades\DB;
use Validator;

class Product extends Model
{
    /** This model uses the SoftDeletes trait for a deleted_at datetime column. */
    use SoftDeletes;

    /** 
     * This models immutable values.
     *
     * @var array 
     */
    protected $guarded = [];

    /** 
     * This models immutable date values.
     * 
     * @var array
     */
    protected $dates = ['deleted_at'];

    /**
     * Set a publicily accessible identifier to get the path for this unique instance.
     * 
     * @return  string
     */
    public function getPathAttribute()
    {
        return url('/products/'.$this->attributes['id']);
    }

    /**
     * Set a publicily accessible identifier to get the image path for this unique instance.
     * 
     * @return  string
     */
    public function getImagePathAttribute()
    {
        return $this->attributes['image_path'] ?? '/image/products/default/not-found.jpg';
    }

    /**
     * This model relationship belongs to \App\Company.
     * 
     * @return  \Illuminate\Database\Eloquent\Model
     */
    public function company()
    {
        return $this->belongsTo('App\Company', 'company_id');
    }

    /**
     * This model relationship has many to \App\ProductReview.
     * 
     * @return  \Illuminate\Database\Eloquent\Model
     */
    public function productReview()
    {
        return $this->hasMany('App\ProductReview', 'product_id');
    }

    /**
     * This model relationship has many to \App\OrderHistoryProducts.
     * 
     * @return  \Illuminate\Database\Eloquent\Model
     */
    public function orderHistoryProducts()
    {
        return $this->hasMany('App\OrderHistoryProducts', 'product_id');
    }

    /**
     * This model relationship belongs to \App\User.
     * 
     * @return  \Illuminate\Database\Eloquent\Model
     */
    public function user()
    {
        return $this->belongsTo('App\User', 'user_id');
    }

    /**
     * Return the formatted cost attribute.
     * 
     * @return  string
     */
    public function getFormattedCostAttribute()
    {
        return "£".number_format($this->attributes['cost'], 2);
    }

    /**
     * Return the cost attribute.
     * 
     * @return  string
     */
    public function getCostAttribute()
    {
        return $this->attributes['cost'];
    }

    /**
     * Query products using request params.
     * 
     * @param  \Illuminate\Database\Eloquent\Model  $query
     * @param  \Illuminate\Http\Request             $request
     * @return \Illuminate\Database\Eloquent\Model
     */
    public function scopeGetProducts($query, $request)
    {
        $querySearch = $request->input('query');
        $min         = $request->input('min_price');
        $max         = $request->input('max_price');
        $sort_by     = $request->input('sort_by');

        $query->select('products.id', 'products.name', 'products.user_id', 'products.company_id', 
                       'products.short_description', 'products.long_description', 'products.product_details', 
                       'products.image_path', 'products.cost', 'products.shippable', 'products.free_delivery', 
                       'products.created_at', 'products.updated_at');
        $whereClause = array();

        if(isset($querySearch))
        {
            $querySearch = filter_var($querySearch, FILTER_SANITIZE_STRING);
            array_push($whereClause, [
                'products.name', 'LIKE', "$querySearch%"
            ]);
        }
        if(isset($min))
        {
            $min = filter_var($min, FILTER_SANITIZE_NUMBER_FLOAT);
            array_push($whereClause, [
                'products.cost', '>', $min
            ]);
        }
        if(isset($max))
        {
            $max = filter_var($max, FILTER_SANITIZE_NUMBER_FLOAT);
            array_push($whereClause, [
                'products.cost', '<', $max
            ]);
        }

        if(isset($whereClause)) $query->where($whereClause);

        switch($sort_by)
        {
            case 'pop': // most popular
                $query->leftJoin('order_history_products', 'products.id', '=', 'order_history_products.product_id')
                      ->groupBy('order_history_products.product_id');
            break;
            case 'top': // top rated
                $query->leftJoin('product_reviews', 'products.id', '=', 'product_reviews.product_id')
                      ->withCount(['productReview as review' => function($query) {
                        $query->select(DB::raw('avg(product_reviews.score) as average_rating'));
                      }])->groupBy('product_reviews.product_id')->orderByDesc('review');
            break;
            case 'low': // lowest price
                $query->orderBy('cost', 'ASC');
            break;
            case 'hig': // highest price
                $query->orderBy('cost', 'DESC');
            break;
            default:
            break;
        }

        $query->orderBy('products.id', 'DESC')->distinct();

        return $query;
    }

    /**
     * Find whether a given user has purchased this product instance.
     * 
     * @param  \App\User  $userId
     * @return bool
     */
    public function didUserPurchaseProduct($userId)
    {
        foreach($this->orderHistoryProducts()->get() as $product){
            $orderHistory = $product->orderHistory()->get();

            foreach($orderHistory as $order)
            {
                if($order->user_id == $userId)
                {
                    return TRUE;
                }
            }
        }

        return FALSE;
    }

    /**
     * Find whether a given user has reviewed this product instance.
     * 
     * @param  \App\User  $userId
     * @return bool
     */
    public function didUserReviewProduct($userId)
    {
        foreach($this->productReview()->get() as $review){

            if($review->user_id == $userId)
            {
                return TRUE;
            }
        }

        return FALSE;
    }

    /**
     * Return the review attribute.
     * 
     * @return  string
     */
    public function getReviewAttribute()
    {
        $review = \App\ProductReview::select(DB::raw('avg(score) as review'))
            ->where('product_id', $this->attributes['id'])
            ->groupBy('product_id')
            ->distinct()->first();

        return isset($review->review) ? number_format((float)$review->review, 2, '.', '') : '0.00';
    }

    /**
     * Get products that belong to a given vendor.
     *
     * @param  \Illuminate\Database\Eloquent\Model  $query, int  $companyId
     * @return \Illuminate\Database\Eloquent\Model
     */
    public function scopeGetCompanyProducts($query, $companyId)
    {
        return $query->where('company_id', '=', $companyId);
    }

    /**
     * Returns boolean indicating if authenticated user owns the current instance of this model.
     *
     * @return bool
     */
    public function doesUserOwnProduct()
    {
        return $this->company->user_id === auth()->user()->id;
    }

    /**
     * Returns boolean indicating whether this model relationship is using a default image.
     *
     * @return bool
     */
    public function usingDefaultImage()
    {
        return $this->attributes['image_path'] === NULL;
    }

    /**
     * Returns an array of errors for \App\Http\Controllers\CompanyProductController requests.
     *
     * @param  \Illuminate\Database\Eloquent\Model  $request
     * @return array
     */
    public static function getErrors($request)
    {
        $validator = Validator::make($request->all(), [
            'name' => 'required|max:191',
            'cost' => 'required|regex:/\d.\d/',
            'shippable' => 'required|boolean|between:0,1',
            'free_delivery' => 'required|boolean|between:0,1',
            'use_default_image' => 'required|boolean|between:0,1',
            'image' => 'image',
            'short_description' => 'required|max:191',
            'long_description' => 'required',
            'product_details' => 'required',
        ]);
        $errors = $validator->errors()->all();

        /** If user doesnt want to use a default image but has not uploaded an image */
        if((bool) $request->input('use_default_image') === FALSE && Input::hasFile('image') === FALSE)
        {
            $errors[] = 'You have opted to not use a default image but you have not provided one.';
        }

        /** If user wants to use default image but has uploaded an image anyway */
        if((bool) $request->input('use_default_image') === TRUE && Input::hasFile('image') === TRUE)
        {
            $errors[] = 'You have opted to use a default image but you provided one anyway.';
        }

        return $errors;
    }
}

File: models/product/product.go
package product

import (
	"fmt"
	"math"
	mathrand "math/rand"
	"strconv"
	"strings"

	"github.com/kkamara/go-ecommerce/config"
	"github.com/kkamara/go-ecommerce/models/company"
	"github.com/kkamara/go-ecommerce/models/helper/number"
	"github.com/kkamara/go-ecommerce/models/helper/pagination"
	"github.com/kkamara/go-ecommerce/models/helper/time"
	"github.com/kkamara/go-ecommerce/models/user"
	"github.com/kkamara/go-ecommerce/schemas"
	"gorm.io/gorm"
	"syreclabs.com/go/faker"
)

func Create(newProduct *schemas.Product) (product *schemas.Product, err error) {
	db, err := config.OpenDB()
	if nil != err {
		return
	}
	now := time.Now()
	newProduct.CreatedAt = now
	newProduct.UpdatedAt = now
	res := db.Create(&newProduct)
	product = newProduct
	if err = res.Error; err != nil {
		return
	}
	return
}

func GetProducts(page, page_size int) (
	products []*schemas.Product,
	pageNum int,
	pageCount int64,
	err error,
) {
	pageNum = page
	db, err := config.OpenDB()
	if err != nil {
		return
	}
	var toCeil float64 = float64(pageCount) / float64(page_size)
	pageCount = int64(math.Ceil(toCeil))
	db.Scopes(pagination.Paginate(page, page_size)).Select(
		`products.id, products.name, products.user_id, products.company_id,
		products.short_description, products.long_description, products.product_details,
		products.image_path, products.cost, products.shippable, products.free_delivery,
		products.created_at, products.updated_at`,
	).Where(
		"products.deleted_at = ?", "",
	).Order(
		"products.id",
	).Find(&products).Count(&pageCount)
	return
}

func Count() (count int64, err error) {
	db, err := config.OpenDB()
	if err != nil {
		return
	}
	var product *schemas.Product
	db.Find(&product).Count(&count)
	return
}

func SearchProducts(params map[string]string, page, page_size int) (
	products []*schemas.Product,
	pageNum int,
	pageCount int64,
	totalRecords int64,
	err error,
) {
	var toCeil float64 = float64(pageCount) / float64(page_size)
	pageCount = int64(math.Ceil(toCeil))
	pageNum = page
	db, err := config.OpenDB()
	if err != nil {
		return
	}

	totalRecords, err = Count()
	if err != nil {
		return
	}

	productSelect := `products.id, products.name, products.user_id, products.company_id,
	products.short_description, products.long_description, products.product_details,
	products.image_path, products.cost, products.shippable, products.free_delivery,
	products.created_at, products.updated_at`
	var q *gorm.DB = db.Scopes(
		pagination.Paginate(page, page_size),
	).Select(productSelect)

	if params["query"] != "" {
		q.Where(
			"products.name LIKE ?",
			fmt.Sprintf("%%%s%%", params["query"]),
		)
	}
	if params["min"] != "" {
		var minFloat float64
		minFloat, err = strconv.ParseFloat(params["min"], 64)
		if err != nil {
			return
		}
		minFloat *= 100
		var min = uint64(minFloat)
		q.Where("products.cost > ?", min)
	}
	if params["max"] != "" {
		var maxFloat float64
		maxFloat, err = strconv.ParseFloat(params["max"], 64)
		if err != nil {
			return
		}
		maxFloat *= 100
		var max = uint64(maxFloat)
		q.Where("products.cost < ?", max)
	}

	switch params["sort_by"] {
	case "pop":
		q.Joins(
			"left join order_products on products.id = order_products.product_id",
		).Group("order_products.product_id")
	case "top":
		reviewSelect := "avg(product_reviews.score) as review"
		q.Select(fmt.Sprintf("%s, %s", productSelect, reviewSelect)).Joins(
			"left join product_reviews on products.id = product_reviews.product_id",
		).Group("product_reviews.product_id").Order("review DESC")
	case "low":
		q.Order("products.cost ASC")
	case "hig":
		q.Order("products.cost DESC")
	case "":
		break
	}

	q.Where(
		"products.deleted_at = ?", "",
	).Order(
		"products.id",
	).Find(&products).Count(&pageCount)
	return
}

func Random() (product *schemas.Product, err error) {
	db, err := config.OpenDB()
	if err != nil {
		return
	}
	var (
		count int64
		c     *schemas.Company
		u     *schemas.User
	)
	res := db.Limit(1).Find(&product).Count(&count)
	if err = res.Error; err != nil {
		return
	}
	if count == 0 {
		c, err = company.Random()
		if err != nil {
			return
		}
		u, err = user.Random("vendor")
		if err != nil {
			return
		}

		now := time.Now()
		cost, _ := number.GetRandomCost()
		newProduct := &schemas.Product{
			UserId:           u.Id,
			CompanyId:        c.Id,
			Name:             faker.Commerce().ProductName(),
			ShortDescription: strings.Join(faker.Lorem().Paragraphs(1), ""),
			LongDescription:  "",
			ProductDetails:   strings.Join(faker.Lorem().Paragraphs(3), "\n\n"),
			ImagePath:        "img/not-found.jpg",
			Cost:             cost,
			Shippable:        mathrand.Intn(2) == 1,
			FreeDelivery:     mathrand.Intn(2) == 1,
			CreatedAt:        now,
			UpdatedAt:        now,
			DeletedAt:        "",
		}

		product, err = Create(newProduct)
		if err != nil {
			return
		}
	}
	return
}

func Seed() (err error) {
	var (
		c *schemas.Company
		u *schemas.User
	)

	for count := 0; count < 30; count++ {
		c, err = company.Random()
		if err != nil {
			return
		}
		u, err = user.Random("vendor")
		if err != nil {
			return
		}

		now := time.Now()
		cost, _ := number.GetRandomCost()
		product := &schemas.Product{
			UserId:           u.Id,
			CompanyId:        c.Id,
			Name:             faker.Commerce().ProductName(),
			ShortDescription: strings.Join(faker.Lorem().Paragraphs(1), ""),
			LongDescription:  "",
			ProductDetails:   strings.Join(faker.Lorem().Paragraphs(3), "\n\n"),
			ImagePath:        "img/not-found.jpg",
			Cost:             cost,
			Shippable:        mathrand.Intn(2) == 1,
			FreeDelivery:     mathrand.Intn(2) == 1,
			CreatedAt:        now,
			UpdatedAt:        now,
			DeletedAt:        "",
		}

		_, err = Create(product)
		if err != nil {
			return
		}
	}
	return
}
